### PR TITLE
[FIX] fields: throw truthy values on old api field

### PR DESCRIPTION
**Pull request by nle-odoo**
[FIX] fields: throw truthy values on old api field
In 7.0, a field overriding an inherited model would overwrite all the
previously set field attributes. In v8 the new API allow us to keep
previous attribute, and only overwrite attributes of our wish.

Following the commit 7b1ef70 (october 2014) an old api field overriding
could conserve previous attributes values in some cases (if the new value
is falsy (None, "", 0, False, (), {}, [], ...) when it should not. It
was partly an optimization to diminish the size of orm registries
dictionaries to save memory (this patch has been tested with all odoo
modules and added +/- 3M).

This commit (proposed by rco) should overwrite falsy value (so the
behavior of v7 for old api fields overwritting is re-introduced) and
get back the behavior of old api in V7.

closes #7035
opw-639712
## Note

To see the impact, I tried to : 
- install all modules installed on runbot,
- check the memory consumption without opening any page and `workers=1`,
- check the memory consumption after opening only one `web`,
- check the memory consumption after opening `web` lots of time,
### without the commit:

| Memory without opening any page | One page opened | Lots of page opened |
| --- | --- | --- |
| 95M resident (267M shared) | 121M resident (287M shared) | 139M (308M) |
| 95M resident (267M shared) | 119M resident (287M shared) | 140M (310M) |
### with this commit:

| Memory without opening any page | One page opened | Lots of page opened |
| --- | --- | --- |
| 98M resident (270M shared) | 122M resident (289M shared) | 143M (312M) |
| 98M resident (270M shared) | 121M resident (289M shared) | 146M (315M) |
### with this commit but without `self._attrs = EMPTY_DICT` :

| Memory without opening any page | One page opened | Lots of page opened |
| --- | --- | --- |
| 97M resident (270M shared) | 122M resident (289M shared) | 146M (315M) |
| 98M resident (270M shared) | 122M resident (289M shared) | 146M (315M) |
## conclusion

seems there may be 3M of difference, also here is a diff of ir_model_fields before and after this commit:

``` diff
--- ofn0.csv    2015-06-11 11:17:14.130069619 +0200
+++ ofn2.csv    2015-06-11 11:15:07.640062888 +0200
@@ -1242,7 +1242,7 @@
 account.invoice,portal_payment_options,,0,Portal Payment Options,html,base,,f,,,,,,,t,,f,f
 account.invoice,reconciled,,0,Paid/Reconciled,boolean,base,,f,,,,,,,t,,t,f
 account.invoice,reference,,0,Communication,char,base,,f,,,,,,,f,,t,f
-account.invoice,reference_type,,0,Communication Type,selection,base,,f,,,,,,,t,,t,t
+account.invoice,reference_type,,0,Communication Type,selection,base,,f,,,,,,,f,,t,t
 account.invoice,residual,,0,Balance,float,base,,f,,,,,,,t,,t,f
 account.invoice,section_id,crm.case.section,0,Sales Team,many2one,base,,f,,,,,,,f,,t,f
 account.invoice,sent,,0,Sent,boolean,base,,f,,,,,,,t,,t,f
@@ -2049,7 +2049,7 @@
 account.voucher,move_ids,account.move.line,0,Journal Items,one2many,base,,f,,,,,,,t,,f,f
 account.voucher,name,,0,Memo,char,base,,f,,,,,,,t,,t,f
 account.voucher,narration,,0,Notes,text,base,,f,,,,,,,t,,t,f
-account.voucher,number,,0,Number,char,base,,f,,,,,,,t,,t,f
+account.voucher,number,,0,Number,char,base,,f,,,,,,,f,,t,f
 account.voucher,paid,,0,Paid,boolean,base,,f,,,,,,,t,,f,f
 account.voucher,paid_amount_in_company_currency,,0,Paid Amount in Company Currency,float,base,,f,,,,,,,t,,f,f
 account.voucher,partner_id,res.partner,0,Partner,many2one,base,,f,,,,,,,t,,t,f
@@ -3135,7 +3135,7 @@
 crm.lead2opportunity.partner.mass,name,,0,Conversion Action,selection,base,,f,,,,,,,,,,
 crm.lead2opportunity.partner.mass,opportunity_ids,crm.lead,0,Opportunities,many2many,base,,f,,,,,,,,,,
 crm.lead2opportunity.partner.mass,partner_id,res.partner,0,Customer,many2one,base,,f,,,,,,,,,,
-crm.lead2opportunity.partner.mass,section_id,crm.case.section,1,Sales Team,many2one,base,,f,,,,,,,,,,
+crm.lead2opportunity.partner.mass,section_id,crm.case.section,0,Sales Team,many2one,base,,f,,,,,,,,,,
 crm.lead2opportunity.partner.mass,user_id,res.users,1,Salesperson,many2one,base,,f,,,,,,,,,,
 crm.lead2opportunity.partner.mass,user_ids,res.users,0,Salesmen,many2many,base,,f,,,,,,,,,,
 crm.lead2opportunity.partner.mass,write_date,,0,Last Updated on,datetime,base,,f,,,,,,,,,,
@@ -4436,7 +4436,7 @@
 gamification.badge.user.wizard,employee_id,hr.employee,0,Employee,many2one,base,,f,,,,,,,,,,
 gamification.badge.user.wizard,id,,0,ID,integer,base,,f,,,,,,,t,,t,f
 gamification.badge.user.wizard,__last_update,,0,Last Modified on,datetime,base,,f,,,,,,,t,,f,f
-gamification.badge.user.wizard,user_id,res.users,0,User,many2one,base,,f,,,,,,,f,,t,t
+gamification.badge.user.wizard,user_id,res.users,0,User,many2one,base,,f,,,,,,,f,,t,f
 gamification.badge.user.wizard,write_date,,0,Last Updated on,datetime,base,,f,,,,,,,f,,t,f
 gamification.badge.user.wizard,write_uid,res.users,0,Last Updated by,many2one,base,,f,,,,,,,f,,t,f
 gamification.challenge,category,,0,Appears in,selection,base,,f,,,,,,,f,,t,t
@@ -10399,7 +10399,7 @@
 res.partner.activation,sequence,,0,Sequence,integer,base,,f,,,,,,,,,,
 res.partner.activation,write_date,,0,Last Updated on,datetime,base,,f,,,,,,,,,,
 res.partner.activation,write_uid,res.users,0,Last Updated by,many2one,base,,f,,,,,,,,,,
-res.partner.bank,acc_number,,0,Account Number,char,base,,f,,,,,,,f,,t,t
+res.partner.bank,acc_number,,0,Account Number,char,base,,f,,,,,,,f,,t,f
 res.partner.bank,bank,res.bank,0,Bank,many2one,base,,f,,,,,,,f,,t,f
 res.partner.bank,bank_bic,,0,Bank Identifier Code,char,base,,f,,,,,,,f,,t,f
 res.partner.bank,bank_code,,0,Bank Code,char,base,,f,,,,,,,,,,
```

so in summary:

2 readonly from true to false:
- account.invoice -> reference_type
- account.voucher -> number

1 select_level from 1 to 0:
- crm.lead2opportunity.partner.mass -> section_id ([old_api](https://github.com/odoo/odoo/blob/8.0/addons/crm/wizard/crm_lead_to_opportunity.py#L191) overwriting [old_api](https://github.com/odoo/odoo/blob/8.0/addons/crm/wizard/crm_lead_to_opportunity.py#L27))

2 required from true to false:
- gamification.badge.user.wizard -> user_id ([old_api](https://github.com/odoo/odoo/blob/8.0/addons/gamification/wizard/grant_badge.py#L31) overwriting [old_api](https://github.com/odoo/odoo/blob/8.0/addons/hr_gamification/wizard/grant_badge.py#L31))
- res.partner.bank -> acc_number

### CODE REVIEW

**Comment by nle-odoo** [comment]
@beledouxdenis 

the first commit contain the change of rco (which looks ok to him), in the second one I checked where this change could impact existing module:

> ### [FIX] fields: correct sides effect of change to old api
The commit 740b7cb get back the behaviour of 7.0 (which changed in
october 2014) that when a field is overrided via the old api, attributes
of a previous definition of the field are lost.

> This fix corrects the case when this behaviour might have had an effect
so specified attributes overrided don't get lost.

> related to opw-639712

issue that could happen is if since november 2014 people overrided fields via the old api in v8+ and used the fact that a `required=True` (or others attributes)  would not be overwrote if a new definition of this field did `required=False` or did not specify `required`.

so with this fix:

expression | before | after
------------ | ------------- | -------------
`'x': fields.char('Hello', required=True),`<br>overwritten by:<br>`'x': fields.char('Hello', required=False),` | True (bad) | False
`'x': fields.char('Hello', required=True),`<br>overwritten by:<br>`'x': fields.char('Hello'),` | True (bad) | False
`'x': fields.char('Hello', required=True),`<br>overwritten by:<br>`'x': fields.char('Hello', required=True),` | True | True
`'x': fields.char('Hello', required=False),`<br>overwritten by:<br>`'x': fields.char('Hello', required=True),` | True | True

**Comment by beledouxdenis** [comment]
I guess this isn't possible to preserve the attributes in the base field if it's not re-defined in the override ?

**Comment by nle-odoo** [comment]
it could be possible (changing the functions signatures and don't set default value but get `**kwargs` and do it), but that would change the behavior the old api always had and changes functions signatures in stable.

**Comment by beledouxdenis** [comment]
Maybe but in the mean time it solves the issue of the customer and won't break the retro-compatibility of existing custom modules relying on the current 8.0 behavior. I would therefore do that if this is possible.

**Comment by beledouxdenis** [comment]
(very nice PR BTW, greatly appreciated :p)

**Comment by nle-odoo** [comment]
the current 8.0 behavior was introduced on 27th october 2014.

* https://github.com/odoo/odoo/pull/7035/files#diff-d883e042eb3efbb47fa00711a76a3409R41 was done without readonly in 2011
* https://github.com/odoo/odoo/pull/7035/files#diff-4a0015012f808c98c6fd27f34572ed2eR191 was done without select=True in 2011
* https://github.com/odoo/odoo/pull/7035/files#diff-7f09b4e70e13750449c3c571dc04f6fcL33 was done without required in 2013
* https://github.com/odoo/odoo/pull/7035/files#diff-02b023e2cdd19e59f7d681db6b0c63a0L227 was done without readonly in 2011

so my guess is that there was an error in those 5 files from the creation (or the original field has been modified but the overrided field have not been changed to take this into account).

so : 

* the first commit gets the behaviour back from {start of time}-november 2014
* the second commit corrects 5 cases which were wrong since 2011-2013 (it's not really related to this commit, it's is only that I stumbled uppon them and that people who installed these addons with updated source from novembler 2014 till now gets the right value thanks to a wrong behaviour)

so in my opinion, the first commit should be used, and the second one is facultative (it's why I split them)

**Comment by nle-odoo** [comment]
and there was another case touched by this commit for `acc_number` it is required by default (see [the code](https://github.com/odoo/odoo/blob/8.0/openerp/addons/base/res/res_bank.py#L117)) and the module `l10n_fr_rib` explicitely `required=False` it:
 https://github.com/odoo/odoo/blob/8.0/addons/l10n_fr_rib/bank.py#L72

but from november 2014 to now it didn't have an effect.
